src/transformers/data/datasets/squad.py

```diff
@@ -128,28 +129,35 @@ def __init__(
         with FileLock(lock_path):
             if os.path.exists(cached_features_file) and not args.overwrite_cache:
                 start = time.time()
-                self.features = torch.load(cached_features_file)
+                cache_dict = torch.load(cached_features_file)
+                self.features = cache_dict["features"]
+                self.dataset = cache_dict["dataset"]
+                self.examples = cache_dict["examples"]
                 logger.info(
                     f"Loading features from cached file {cached_features_file} [took %.3f s]", time.time() - start
                 )
             else:
                 if mode == Split.dev:
-                    examples = self.processor.get_dev_examples(args.data_dir)
+                    self.examples = self.processor.get_dev_examples(args.data_dir)
                 else:
-                    examples = self.processor.get_train_examples(args.data_dir)
+                    self.examples = self.processor.get_train_examples(args.data_dir)
 
-                self.features = squad_convert_examples_to_features(
-                    examples=examples,
+                self.features, self.dataset = squad_convert_examples_to_features(
+                    examples=self.examples,
                     tokenizer=tokenizer,
                     max_seq_length=args.max_seq_length,
                     doc_stride=args.doc_stride,
                     max_query_length=args.max_query_length,
                     is_training=mode == Split.train,
                     threads=args.threads,
+                    return_dataset=dataset_format,
                 )
 
                 start = time.time()
-                torch.save(self.features, cached_features_file)
+                torch.save(
```

Comment:
Do this save if the attrs is not None, or else use `self.old_features` to keep backward compatibility.

Reply:
@sgugger I'm sorry, I don't think I fully understand this suggestion. As I see it, this branch of the code will only execute if there was no cache file, so there will be no `self.old_features` defined and the code should be able to save the new cache file format since it re-derived everything.

Reply:
Good point.